lib/widgets/compose_box.dart

```diff
@@ -44,21 +47,75 @@ class TopicTextEditingController extends TextEditingController {
 
 enum ContentValidationError {
   empty,
-  tooLong;
+  tooLong,
+  uploadInProgress;
 
-  // Later: upload in progress; quote-and-reply in progress
+  // Later: quote-and-reply in progress
 
   String message() {
     switch (this) {
       case ContentValidationError.tooLong:
         return "Message length shouldn't be greater than 10000 characters.";
       case ContentValidationError.empty:
         return 'You have nothing to send!';
+      case ContentValidationError.uploadInProgress:
+        return 'Please wait for the upload to complete.';
     }
   }
 }
 
 class ContentTextEditingController extends TextEditingController {
+  int _uploadCounter = 0;
+
+  final Map<int, ({String filename, String placeholder})> _uploads = {};
+
+  /// The cursor position, if selection is valid and collapsed, else text.length.
+  TextRange _cursorPositionOrEnd() {
+    final TextRange selection = value.selection;
+    final String text = value.text;
+    final int index = (selection.isValid && selection.end < text.length && selection.isCollapsed)
+      ? selection.end
+      : text.length;
+    return TextRange.collapsed(index);
+  }
+
+  /// Tells the controller that a file upload has started.
+  ///
+  /// Returns an int "tag" that should be passed to registerUploadEnd on the
+  /// upload's success or failure.
+  int registerUploadStart(String filename) {
+    final tag = _uploadCounter;
+    _uploadCounter += 1;
+    final placeholder = '[Uploading $filename...]()\n\n'; // TODO(i18n)
+    _uploads[tag] = (filename: filename, placeholder: placeholder);
+    notifyListeners(); // _uploads change could affect validationErrors
+    value = value.replaced(_cursorPositionOrEnd(), placeholder);
```

Comment:
`zulip-mobile` differs in the case where the selection is not collapsed, i.e., instead of a blinking cursor, there's a highlight on some of the text.

We still don't replace the selected text; that's unchanged here. But:

- In `zulip-mobile`, selection-end is taken to be the cursor position, and the placeholder is inserted there.
- In this branch, the placeholder is inserted at the end of the whole content-input text.

Reply:
I wouldn't mind following `zulip-mobile` here, but here's an attempt to describe why I was drawn to this new approach:

I think users are unlikely to see text selection as one of the steps to take to attach a file. The `zulip-mobile` design follows from that judgment too: there, instead of (e.g.) replacing the selected text, we ignore the fact that some text was selected and we insert the markdown at selection-end.

Then I thought, if some text is selected, it's probably from some previous interaction, like looking up a word in the dictionary. Or it's just an accident. Anyway, the user probably doesn't know (or has forgotten) that the file-upload UI wants a hint for a single index at which the user will happily see the markdown inserted. So I figured nothing in `selection` (like the start or end) will predict that index well, and we might as well do what we do when the input is unfocused, and what "attach" normally means, which is to put it at the end.

Reply:
We discussed briefly in the office just now. We'll go and do the insertion where the selection is, not because that's a spot the user was probably asking for it to go exactly (if they did have an exact spot in mind that they were trying to convey, they'd likely have placed a cursor and not a whole selection range), but because we want it to be in a place they're paying attention to, and the selection range will be that.

In particular, if the draft content is long, the end of the content may not even be visible on screen — if we insert there, the user may perceive that nothing happened. The selection area will be on screen, and it has a conspicuous pair of selection handles and conspicuous highlighting.